### DISCUSSION TITLE
File chooser dialog can only be shown with a user activation.

### DISCUSSION THREAD

**Discussion by phoducgiang** [cefsharp/CefSharp#3710] [General]
I have to embedded cefsharp to my winforms tools. However i cannot run the js script for click to show dialog box then choose the file for upload. Cefsharp always show the log "File chooser dialog can only be shown with a user activation.", please advise me how to fix this issue.

Here is my js script:
`function performClick(elemId) {
   var elem = document.getElementsByName(elemId)[0];
   if(elem && document.createEvent) {
      var evt = document.createEvent("MouseEvents");
      evt.initEvent("click", true, false);
      elem.dispatchEvent(evt);
   }
}
performClick("tabSearch$TabPanel1$fileSearchTemplate");`

![Capture](https://user-images.githubusercontent.com/80267275/126689911-72ee3729-2abc-4ee2-bef2-3fbf9ec6c0bc.PNG)

**Comment by amaitland** [comment]
For security reasons you cannot trigger a file dialog through JavaScript without user action. You can focus the button and send through an enter key press.

https://stackoverflow.com/questions/44736904/issue-with-cefsharp-browser-sendkeys

**Reply by amaitland** [reply]
You might also be able to execute your JavaScript through the DevTools protocol. 

https://chromedevtools.github.io/devtools-protocol/

https://github.com/cefsharp/CefSharp/issues/3165

**Reply by phoducgiang** [reply]
The sendkey event working on Winforms ?. I have to tried with solution in stackoverflow site provided and seem the "enter" button not send to cefsharp.

**Reply by amaitland** [reply]
You need to make sure the ChromiumWebBrowser has focus.

**Reply by phoducgiang** [reply]
while the running code the ChromiumWebBrowser has focus and i did not see any response from sendkey event. Howerver the Log from debug still "File chooser dialog can only be shown with a user activation."


My C# code:
```c#         ExcuteScript("document.getElementById(\"pnlSearch\").click()");
                ExcuteScript("document.getElementById(\"tabSearch_TabPanel1_Label12\").click()");
                ExcuteScript("document.getElementsByName(\"tabSearch$TabPanel1$fileSearchTemplate\")[0].focus()");
                KeyEvent k = new KeyEvent();
                k.WindowsKeyCode = 0x0D;
                k.FocusOnEditableField = true;
                k.IsSystemKey = false;
                k.Type = KeyEventType.Char;
                chromeBrowser.Invoke(new Action(() =>
                {
                    chromeBrowser.Focus();
                    chromeBrowser.GetBrowser().GetHost().SendKeyEvent(k);
                }));
```

**Reply by amaitland** [reply]
Give the ChromiumWebBrowser instance focus before executing your JavaScript. 

Send a proper key press, both down and up events.

https://stackoverflow.com/a/53452094/4583726

**Reply by phoducgiang** [reply]
@amaitland I have to tried with your instruction and still no response.

can you try to sendkeys event button for website

I want to press the choose file button.
![Capture](https://user-images.githubusercontent.com/80267275/126735968-95d0dae9-c5fa-414e-918f-6b4ae866d344.PNG)

**Reply by amaitland** [reply]
If your intention is to programatically upload a file then the DevTools protocol is better suited as you can directly set the file.

https://stackoverflow.com/a/65625960/4583726

**Reply by phoducgiang** [reply]
Manually choose file and upload is ok but the programatically upload file is error warning from this site :(

**Reply by amaitland** [reply]
Did you try the DOM.SetFileInputFiles approach via the DevTools protocol?

**Reply by phoducgiang** [reply]
@amaitland  ofcourse bro. The website response as attached image
![Capture](https://user-images.githubusercontent.com/80267275/126740614-da276e09-062c-4a0a-b2e6-a06418d8cc15.PNG)

**Reply by amaitland** [reply]
I'd suggest you try a different website for testing purposes.

**Reply by amaitland** [reply]
You can open the DevTools interface to inspect the request, you should compare a working and non working to see what's going on. 

https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

**Reply by ukandrewc** [reply]
I had been working on something similar and can confirm that using DevTools Protocol DOM.SetFileInputFiles does work.

This is using a DOM wrapper in VB, so you'll need to adapt it.
```
'Get File Input
Dim Elm = CSView.document.evaluate("//input[@type='file']")
'Visual check
Elm.style("background") = "red"
'Get location to find backendNodeId
Dim Rec = Elm.getBoundingClientRect
Dim Res = Await CSView.DevTools.DOM.GetNodeForLocationAsync(Rec.X, Rec.Y)
Dim Inp = Await CSView.DevTools.DOM.SetFileInputFilesAsync({"c:\myfolder\myfile.pdf"}, backendNodeId:=Res.BackendNodeId)
Elm.nextElementSibling.click()
```

**Reply by phoducgiang** [reply]
@amaitland can you shown about getBoundingClientRect with winform ?
@ukandrewc i wil try it

**Reply by amaitland** [reply]
> can you shown about getBoundingClientRect with winform ?

Personally this is not the approach I would choose. 

A super simple example looks like this:

```c#
using (var devTools = browser.GetDevToolsClient())
{
	//Get Document
	var document = await devTools.DOM.GetDocumentAsync();

	//Get the NodeId for our File Upload element based on it's element Id
	var inputFileNode = await devTools.DOM.QuerySelectorAsync(document.Root.NodeId, "#myFile");
	//Upload a file
	var setFileInputResponse = await devTools.DOM.SetFileInputFilesAsync(new string[] { @"C:\Windows\win.ini" }, inputFileNode.NodeId);
}
```

An example of uploading a file to https://www.w3schools.com/howto/tryit.asp?filename=tryhow_html_file_upload_button which uses an Iframe looks like:

```c#
using (var devTools = browser.GetDevToolsClient())
{
	//Get Document
	var document = await devTools.DOM.GetDocumentAsync();

	//Find our iFrame
	var iFrameNode = await devTools.DOM.QuerySelectorAsync(document.Root.NodeId, "#iframeResult");
	//Get more detail of around iFrame
	var iFrameNodeDetail = await devTools.DOM.DescribeNodeAsync(iFrameNode.NodeId, depth: -1, pierce: true);
	//Get the NodeId for the BackendNodeId
	var convertedNodeIds = await devTools.DOM.PushNodesByBackendIdsToFrontendAsync(new[] { iFrameNodeDetail.Node.ContentDocument.BackendNodeId });
	
	//Get the input element Node Id based on it's element Id
	var inputFileNode = await devTools.DOM.QuerySelectorAsync(convertedNodeIds.NodeIds[0], "#myFile");

	//Upload file
	var setFileInputResponse = await devTools.DOM.SetFileInputFilesAsync(new string[] { @"C:\Windows\win.ini" }, inputFileNode.NodeId);
}
```

How you get the `NodeId` for the input element depends on your `HTML`. You should be able to use the query selector synatx with something like `input[name='tabSearch$TabPanel1$fileSearchTemplate']`

```c#
var inputFileNode = await devTools.DOM.QuerySelectorAsync(document.Root.NodeId, "input[name='tabSearch$TabPanel1$fileSearchTemplate']");
```

I've not tested your the last part. You should read https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

**Reply by phoducgiang** [reply]
@amaitland I don't know why when I tried your example code with my case, the exception showed as attached image. However when I using devtools for queryselector, I got this Node Id.
![Capture](https://user-images.githubusercontent.com/80267275/126903557-801af73e-1ae6-498f-a7cf-36a3c61aab0a.PNG)

**Reply by amaitland** [reply]
Please no images of exceptions, use the copy details button and provide the error as text in future. 

You'll need to update the QuerySelectorAsync to find the DOM node for particular site.

Read https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

You can use the DevTools console to test out your query.
